### PR TITLE
Color Space management

### PR DESCRIPTION
**Pull request by demoulinv**
## Description
A new private metadata named "AliceVision:ColorSpace" is now managed.
Read and Write image update to enable image processing in a color space given as a parameter.
imageProcessing executable updated to take advantage of the above possibility.

## Features list
- [X] AliceVision:ColorSpace metadata added to image file when writing. Possible values are in the list srgb, srgb_linear, aces, acescg.
- [X] Enable image processing in a color space given as a parameter.
- [X] Implement color space conversion with OCIO at image reading to enable the selection of another color space than linear or sRGB as working color space.
- [X] Check existence of the OCIO config file and throw an error if not found when needed. 3 environment variables are considered, ALICEVISION_OCIO, OCIO and ALICEVISION_ROOT, see implementation details.
- [X] Set the working color space as a parameter in the imageProcessing executable.

## Implementation remarks
- Function `getColorConfigFilePath()` defined for avoiding some code duplication.
- OCIO Color config file is first searched by using the ALICEVISION_OCIO env var. If it is undefined or points an unexisting file, the standardized OCIO env var is considered. If it is undefined or points an unexisting file, the ALICEVISION_ROOT env var is considered to point the embedded config file. If the ALICEVISION_OCIO or OCIO environment variables are set, they must be set with the path of the OCIO configuration file to be used.
- Function `getImageColorSpace(imagePath)` defined.
- Color space conversion already implemented at image reading. The target color space, previously named `outputcolorSpace `, field of the `imageReadOptions `structure is renamed in `workingColorSpace`.
- In order to perform the rigth conversion, the image writer check the AliceVision:ColorSpace metadata, set when processing is done, to identify the working color space.
- For clarification the parameter `imageColorSpace `of the `writeImage `methods is renamed `outputImageColorSpace`.

### CODE REVIEW

File: src/aliceVision/image/io.cpp [hunk 1]
```diff
@@ -92,6 +99,75 @@ std::istream& operator>>(std::istream& in, EImageColorSpace& dataType)
...
+    }
+    else
+    {
+        ALICEVISION_LOG_TRACE("OCIO configuration file: '" << configOCIOFilePath << "' found.");
+    }
+
+    return configOCIOFilePath;
+}
+
+EImageColorSpace getImageColorSpace(const std::string imagePath)
```
Comment (fabiencastan) [inline src/aliceVision/image/io.cpp:124]: const std::string&

Comment (demoulinv) [reply src/aliceVision/image/io.cpp:124]: Does not compile anymore on Windows.

**Comment by fabiencastan** [comment]
I don't understand why we have new build errors on windows CI which does not seem to be really related to the PR:

```
  D:\a\AliceVision\vcpkg\installed\x64-windows-release\include\boost/mpl/aux_/preprocessor/def_params_tail.hpp(23,10): fatal  error C1083: Cannot open include file: 'boost/preprocessor/identity.hpp': No such file or directory (compiling source file D:\a\AliceVision\AliceVision\src\aliceVision\system\Logger.cpp) [D:\a\AliceVision\AliceVision\build\src\aliceVision\system\aliceVision_system.vcxproj]
```

**Comment by p12tic** [comment]
@fabiencastan I think the commit
[SRGB_LINEAR becomes LINEAR to be more compliant with the OCIO config](https://github.com/alicevision/AliceVision/pull/1170/commits/443a848cefa26d9a635b4ef0a84a37d8ec2f3557) should be combined into the parent commit  [[Add AliceVision Color Space metadata]](https://github.com/alicevision/AliceVision/pull/1170/commits/68a0aef67312d5ad85f20ae423cd5d1cc6cbb0d9). The parent commit renames LINEAR to SRGB_LINEAR and the child commit renames the colorspace back. 

There are also a couple of left-overs `EImageColorSpace::SRGB_LINEAR` changes which I think should be fixed. It's interesting why they don't cause compile errors, maybe they are in code that's disabled..

File: src/aliceVision/image/io_test.cpp [hunk 2]
```diff
@@ -22,8 +22,23 @@ using namespace aliceVision;
...
+    if (!overwrite) {
+        size_t envsize = 0;
+        errcode = getenv_s(&envsize, NULL, 0, name);
+        if (errcode || envsize) return errcode;
+    }
+    return _putenv_s(name, value);
+}
+#endif
+
+int err = setenv("ALICEVISION_ROOT", std::string(THIS_SOURCE_DIR).c_str(), 1);
```
Comment (p12tic) [inline src/aliceVision/image/io_test.cpp:38]: I think instead of overriding `ALICEVISION_ROOT` via setenv, we should retrieve the environmental variable via a function in which we could add overriding functionality. Something like:

```
static std::string aliceVisionRootOverride;

std::string getAliceVisionRoot()
{
    if (!aliceVisionRootOverride.empty())
        return aliceVisionRootOverride;
    const char* value = std::getenv("ALICEVISION_ROOT");
    return value ? value : "";
}

void setAliceVisionRootOverride(const std::string& value)
{
    aliceVisionRootOverride = value;
}
```

Comment (p12tic) [reply src/aliceVision/image/io_test.cpp:38]: I've just did a PR with this: https://github.com/alicevision/AliceVision/pull/1268

File: src/aliceVision/hdr/hdr_test.cpp [hunk 3]
```diff
@@ -78,7 +78,7 @@ bool buildBrackets(std::vector<std::string>& paths, std::vector<float>& times, c
 
         ALICEVISION_LOG_INFO("writing to " << temp.string());
 
-        image::writeImage(temp.string(), img_bracket, image::EImageColorSpace::LINEAR);
+        image::writeImage(temp.string(), img_bracket, image::EImageColorSpace::SRGB_LINEAR);
```
Comment (p12tic) [inline src/aliceVision/hdr/hdr_test.cpp:81]: This will break hdr_test.cpp (which is currently disabled, but we shouldn't introduce unnecessary changes that break existing code).

Comment (demoulinv) [reply src/aliceVision/hdr/hdr_test.cpp:81]: This change makes matching AliceVision color space naming with OCIO one. In OCIO "linear" points out the color space defined to be the scene linear one. It could be ACEScg, Rec709 linear, or any other linear color space. This change allows to take advantage on that in AV.

Comment (p12tic) [reply src/aliceVision/hdr/hdr_test.cpp:81]: @demoulinv I understand the intention, but this PR doesn't actually introduce SRGB_LINEAR. It introduced it in the first commit [[Add AliceVision Color Space metadata]](https://github.com/alicevision/AliceVision/pull/1170/commits/c1c77210b695536ed941af543cb323800ebb8f9a), but the second commit [SRGB_LINEAR becomes LINEAR to be more compliant with the OCIO config](https://github.com/alicevision/AliceVision/pull/1170/commits/eeae5f871dc0d009318b7952053f3c5c476d5924) reverted everything back.

File: src/aliceVision/image/imageAlgo.cpp [hunk 4]
```diff
@@ -151,8 +151,7 @@ void colorconvert(oiio::ImageBuf& imgBuf, image::EImageColorSpace fromColorSpace
     {
         if(fromColorSpace == EImageColorSpace::SRGB)
             oiio::ImageBufAlgo::colorconvert(imgBuf, imgBuf,
-                                             EImageColorSpace_enumToOIIOString(EImageColorSpace::SRGB),
-                                             EImageColorSpace_enumToOIIOString(EImageColorSpace::LINEAR));
+                                             EImageColorSpace_enumToString(EImageColorSpace::SRGB), EImageColorSpace_enumToString(EImageColorSpace::LINEAR));
```
[the remark this answers is not in the document]
Comment (p12tic) [reply src/aliceVision/image/imageAlgo.cpp:154]: Well, not break in this particular case, but we should consistently use EImageColorSpace_enumToOIIOString to pass colorspaces to OIIO, because other enum values may be different.

File: src/aliceVision/image/io.cpp [hunk 5]
```diff
@@ -26,15 +26,114 @@
...
+            {
+                ALICEVISION_LOG_TRACE("ALICEVISION_OCIO is empty.");
+            }
+            else
+            {
+                ALICEVISION_LOG_WARNING("ALICEVISION_OCIO is defined but does not point to an existing file: '" << configOCIOFilePath << "'");
+            }
+        }
+
+        // To be uncommented to take OCIO env var in consideration before using the enbedded config file
```
Comment (p12tic) [inline src/aliceVision/image/io.cpp:66]: Commented out code will bitrot. I think it would make sense to put everything in an `if` with condition hardcoded to false.

File: src/aliceVision/image/io.cpp [hunk 6]
```diff
@@ -440,7 +584,13 @@ void readImage(const std::string& path,
   configSpec.attribute("raw:use_camera_wb", (imageReadOptions.applyWhiteBalance?1:0)); // white balance correction
   // use_camera_matrix: Whether to use the embedded color profile, if it is present: 0=never, 1 (default)=only for DNG files, 3=always
   configSpec.attribute("raw:use_camera_matrix", 3); // use embeded color profile
+
+#if OIIO_VERSION <= (10000 * 2 + 100 * 0 + 8) // OIIO_VERSION <= 2.0.8
```
Comment (p12tic) [inline src/aliceVision/image/io.cpp:588]: We have `find_package(OpenImageIO 2.0.9 REQUIRED)` in CMakeLists.txt, so this code is not needed.

File: src/aliceVision/image/io.cpp (hunk 5 above)
Comment (fabiencastan) [reply src/aliceVision/image/io.cpp:66]: ok

File: src/aliceVision/image/imageAlgo.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (fabiencastan) [reply src/aliceVision/image/imageAlgo.cpp:154]: Ouch, good catch!

File: src/aliceVision/image/io.cpp (hunk 6 above)
Comment (fabiencastan) [reply src/aliceVision/image/io.cpp:588]: Yes, looks like an error during the rebase.

**Review by fabiencastan — APPROVED** [review]
